### PR TITLE
Expose BLS functionality trough web-client

### PR DESCRIPTION
**Pull request by viquezclaudio**
This is required by the POS migration tool in order to generate and import BLS keys.
This functionality needs to be compiled and exposed for NodeJS

## Pull request checklist

- [x] All tests pass. The project builds and runs.
- [x] I have resolved any merge conflicts.
- [x] I have resolved all `clippy` and `rustfmt` warnings.

### CODE REVIEW

File: web-client/Cargo.toml [hunk 1]
```diff
@@ -38,6 +38,7 @@ nimiq-account = { path = "../primitives/account", default-features = false }
 nimiq-block = { path = "../primitives/block" }
 nimiq-blockchain-interface = { path = "../blockchain-interface" }
 nimiq-blockchain-proxy = { path = "../blockchain-proxy", default-features = false }
+nimiq-bls = {path = "../bls"}
```
Comment (jsdanielh) [inline web-client/Cargo.toml]:
```suggestion
nimiq-bls = { path = "../bls" }
```

File: web-client/dist/package-lock.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{
```
Comment (jsdanielh) [inline web-client/dist/package-lock.json]: Added file by mistake?

**Review by jsdanielh — CHANGES_REQUESTED** [review]

File: web-client/src/client/mod.rs [hunk 3]
```diff
@@ -0,0 +1,5 @@
+pub mod account;
+pub mod block;
+pub mod client;
```
Comment (jsdanielh) [inline web-client/src/client/mod.rs]: Could you address this clippy warning?

**Review by jsdanielh — CHANGES_REQUESTED** [review]

**Comment by jsdanielh** [comment]
There are several clippy warnings, could you address them?

**Review by jsdanielh — APPROVED** [review]

**Review by sisou — APPROVED** [review]

File: web-client/src/primitives/bls_secret_key.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+#[wasm_bindgen]
+impl BLSSecretKey {
+    /// Generates a new private key from secure randomness.
+    pub fn generate() -> BLSSecretKey {
+        BLSSecretKey::from_native(nimiq_bls::SecretKey::generate_default_csprng())
+    }
+
+    /// Deserializes a private key from a byte array.
+    pub fn unserialize(bytes: &[u8]) -> Result<BLSSecretKey, JsError> {
+        let key = nimiq_bls::SecretKey::deserialize_from_vec(&mut &*bytes)?;
```
Comment (jsdanielh) [inline web-client/src/primitives/bls_secret_key.rs]: Clippy warning, could you fix it?

File: web-client/src/primitives/bls_public_key.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+#[wasm_bindgen]
+impl BLSPublicKey {
+    /// Derives a public key from an existing private key.
+    pub fn derive(secret_key: &BLSSecretKey) -> BLSPublicKey {
+        BLSPublicKey::from_native(nimiq_bls::PublicKey::from_secret(secret_key.native_ref()))
+    }
+
+    /// Deserializes a public key from a byte array.
+    pub fn unserialize(bytes: &[u8]) -> Result<BLSPublicKey, JsError> {
+        let key = nimiq_bls::PublicKey::deserialize_from_vec(&mut &*bytes)?;
```
Comment (jsdanielh) [inline web-client/src/primitives/bls_public_key.rs]: Clippy warning, could you fix it?

File: web-client/src/primitives/bls_key_pair.rs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+    /// Derives a keypair from an existing private key.
+    pub fn derive(private_key: &BLSSecretKey) -> BLSKeyPair {
+        let key_pair = nimiq_bls::KeyPair::from(*private_key.native_ref());
+        BLSKeyPair::from_native(key_pair)
+    }
+
+    /// Deserializes a keypair from a byte array.
+    pub fn unserialize(bytes: &[u8]) -> Result<BLSKeyPair, JsError> {
+        let key_pair = nimiq_bls::KeyPair::deserialize_from_vec(&mut &*bytes)?;
```
Comment (jsdanielh) [inline web-client/src/primitives/bls_key_pair.rs]: Clippy warning, could you fix it?

**Review by jsdanielh — CHANGES_REQUESTED** [review]

**Review by jsdanielh — CHANGES_REQUESTED** [review]
There are still a couple of clippy warnings

File: web-client/src/primitives/bls_key_pair.rs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            secret_key: *secret_key.native_ref(),
+            public_key: *public_key.native_ref(),
+        };
+        BLSKeyPair::from_native(key_pair)
+    }
+
+    /// Serializes to a byte array.
+    pub fn serialize(&self) -> Vec<u8> {
+        let mut vec = self.inner.serialize_to_vec();
+        vec.push(0);
```
Comment (jsdanielh) [inline web-client/src/primitives/bls_key_pair.rs]: Why are we pushing an extra `0u8`?

File: web-client/src/primitives/bls_secret_key.rs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    /// Serializes the private key to a byte array.
+    pub fn serialize(&self) -> Vec<u8> {
+        self.inner.serialize_to_vec()
+    }
+
+    /// Parses a private key from its hex representation.
+    #[wasm_bindgen(js_name = fromHex)]
+    pub fn from_hex(hex: &str) -> Result<BLSSecretKey, JsError> {
+        let raw = hex::decode(hex)?;
+        // 95 and 96 byres are valid sizes.
```
Comment (jsdanielh) [inline web-client/src/primitives/bls_secret_key.rs]:
```suggestion
        // 95 and 96 bytes are valid sizes.
```

Comment (jsdanielh) [inline web-client/src/primitives/bls_secret_key.rs]: 96 bytes are no longer valid

**Review by jsdanielh — CHANGES_REQUESTED** [review]

File: web-client/src/primitives/bls_key_pair.rs [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+            secret_key: *secret_key.native_ref(),
+            public_key: *public_key.native_ref(),
+        };
+        BLSKeyPair::from_native(key_pair)
+    }
+
+    /// Serializes to a byte array.
+    pub fn serialize(&self) -> Vec<u8> {
+        let mut vec = self.inner.serialize_to_vec();
+        vec
```
Comment (jsdanielh) [inline web-client/src/primitives/bls_key_pair.rs]: This would remove the clippy warning:
```suggestion
        self.inner.serialize_to_vec()
```

**Review by jsdanielh — CHANGES_REQUESTED** [review]

**Review by jsdanielh — APPROVED** [review]

**Review by jsdanielh — APPROVED** [review]

**Pull request merged**